r/tests/testthat/test-record-batch-reader.R

```diff
@@ -139,3 +139,21 @@ test_that("reader with 0 batches", {
   expect_r6_class(tab, "Table")
   expect_identical(dim(tab), c(0L, 1L))
 })
+
+test_that("reader head method edge cases", {
+  batch <- record_batch(
+    x = 1:10,
+    y = letters[1:10]
+  )
+  sink <- BufferOutputStream$create()
+  writer <- RecordBatchStreamWriter$create(sink, batch$schema)
+  writer$write(batch)
+  writer$write(batch)
+  writer$close()
+  buf <- sink$finish()
+
+  reader <- RecordBatchStreamReader$create(buf)
+  expect_error(head(reader, -1)) # Not (yet) supported
+  expect_equal(head(reader, 0), Table$create(x = integer(0), y = character(0)))
+  expect_equal(head(reader, 100), Table$create(batch, batch))
```

Comment:
If I'm reading the head method correctly, it reads from the RecordBatchReader, but pushes the records back on so that they can be read again if necessary. Should we add a test that confirms this?

Something like:
```
expect_equal(head(reader, 10), Table$create(batch))
expect_equal(reader$read_table(), Table$create(batch, batch))
```

Reply:
Actually I believe the opposite is true: the method is consuming the reader. RecordBatchReader (and Scanner) are one-shot. (I think you may be misreading the C++ function: `reader` is the reader and we collect into `batches` via `push_back`, there is no pushing back into `reader`.)